File: apps/python/lib/chapter15/service.py
from collections.abc import Callable

from lib.chapter15.domain import (
    ModelNotFoundError,
    ModelStore,
    Movie,
    Passenger,
    SalesPrediction,
    SurvivalPrediction,
)


class PredictionService:
    def __init__(self, store: ModelStore) -> None:
        self.store = store

    def predict_sales(self, movie: Movie) -> SalesPrediction:
        model = self.store.load_sales_model()
        return SalesPrediction(sales=model.predict_sales(movie))

    def predict_survival(self, passenger: Passenger) -> SurvivalPrediction:
        model = self.store.load_survival_model()
        return SurvivalPrediction.from_probability(
            model.survival_probability(passenger)
        )

    def health(self) -> dict[str, bool]:
        return {
            "cinema": self._can_load(self.store.load_sales_model),
            "survived": self._can_load(self.store.load_survival_model),
        }

    @staticmethod
    def _can_load(load: Callable[[], object]) -> bool:
        try:
            load()
        except ModelNotFoundError:
            return False
        return True

File: apps/clojure/src/getting_started_ml/chapter15/service.clj
(ns getting-started-ml.chapter15.service
  "第 15 章のアプリケーション層。置き場からモデルを読み込んで予測する。HTTP を知らない。

   置き場は ModelStore の約束を満たすものなら何でもよいので、テストでは reify の偽物を渡せる。
   状態を持たないので、同時に走るハンドラーから呼ばれても困らない。"
  (:require [getting-started-ml.chapter15.domain :as domain]))

(defn predict-sales
  "映画の特徴量から興行収入を予測する。"
  [store movie]
  ((domain/load-sales-model store) movie))

(defn predict-survival
  "乗客の特徴量から生存するかどうかを予測する。"
  [store passenger]
  ((domain/load-survival-model store) passenger))

(defn- ready?
  "モデルを読み込めるかどうか。読み込めない理由がほかにあれば、そのまま投げる。"
  [load store]
  (try
    (load store)
    true
    (catch clojure.lang.ExceptionInfo e
      (if (domain/model-not-found? e) false (throw e)))))

(defn health
  "モデルごとに、読み込めるかどうかを返す。
   マップは 9 要素以上で順を保たないが、ここは 2 つなので array-map で並びを固定できる。"
  [store]
  (array-map domain/sales-model (ready? domain/load-sales-model store)
             domain/survival-model (ready? domain/load-survival-model store)))
